File: problems/LeetCode/TopInterview150/maxProfit.ts
function maxProfit(prices: number[]): number {
  if (prices.length === 0) return 0;

  let lowest = prices[0];
  let best = 0;

  for (let price of prices) {
    if (price < lowest) {
      lowest = price;
      continue;
    }
    best = Math.max(best, price - lowest);
  }

  return best;
}

File: problems/LeetCode/TopInterview150/maxProfit.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        profit = 0
        lowest = prices[0]

        for i in range(1, len(prices)):
            if prices[i] < lowest:
                lowest = prices[i]
                continue

            profit = max(profit, prices[i] - lowest)

        return profit
        
# Time: O(n)
# Space: O(1)